File: gherkin/java/src/main/java/io/cucumber/gherkin/GherkinLine.java
package io.cucumber.gherkin;

import java.util.ArrayList;
import java.util.List;
import java.util.PrimitiveIterator;
import java.util.Scanner;

import static io.cucumber.gherkin.GherkinLanguageConstants.COMMENT_PREFIX;
import static io.cucumber.gherkin.GherkinLanguageConstants.TAG_PREFIX;
import static io.cucumber.gherkin.StringUtils.ltrim;
import static io.cucumber.gherkin.StringUtils.rtrim;
import static io.cucumber.gherkin.StringUtils.symbolCount;
import static io.cucumber.gherkin.StringUtils.trim;

public class GherkinLine implements IGherkinLine {
    // TODO: set this to 0 when/if we change to 0-indexed columns
    private static final int OFFSET = 1;
    private final String lineText;
    private final String trimmedLineText;
    private final int indent;
    private final int line;

    public GherkinLine(String lineText, int line) {
        this.lineText = lineText;
        this.trimmedLineText = trim(lineText);
        this.line = line;
        indent = symbolCount(lineText) - symbolCount(ltrim(lineText));
    }

    @Override
    public int indent() {
        return indent;
    }

    @Override
    public void detach() {

    }

    @Override
    public String getLineText(int indentToRemove) {
        if (indentToRemove < 0 || indentToRemove > indent())
            return trimmedLineText;
        return lineText.substring(indentToRemove);
    }

    @Override
    public boolean isEmpty() {
        return trimmedLineText.length() == 0;
    }

    @Override
    public boolean startsWith(String prefix) {
        return trimmedLineText.startsWith(prefix);
    }

    @Override
    public String getRestTrimmed(int length) {
        return trimmedLineText.substring(length).trim();
    }

    @Override
    public List<GherkinLineSpan> getTags() {

        String uncommentedLine = trimmedLineText.split("\\s" + COMMENT_PREFIX, 2)[0];
        List<GherkinLineSpan> tags = new ArrayList<>();
        int indexInUncommentedLine = 0;

        String[] elements = uncommentedLine.split(TAG_PREFIX);
        for (String element : elements) {
            String token = rtrim(element);
            if (token.isEmpty()) {
                continue;
            }
            int symbolLength = uncommentedLine.codePointCount(0, indexInUncommentedLine);
            int column = indent() + symbolLength + 1;
            if (!token.matches("^\\S+$")) {
                throw new ParserException("A tag may not contain whitespace: " + trimmedLineText, new Location(line, column));
            }
            tags.add(new GherkinLineSpan(column, TAG_PREFIX + token));
            indexInUncommentedLine += element.length() + 1;
        }
        return tags;
    }

    @Override
    public List<GherkinLineSpan> getTableCells() {
        List<GherkinLineSpan> lineSpans = new ArrayList<>();
        StringBuilder cellBuilder = new StringBuilder();
        boolean beforeFirst = true;
        int col = 0;
        int cellStart = 0;
        boolean escape = false;
        PrimitiveIterator.OfInt iterator = lineText.codePoints().iterator();
        while (iterator.hasNext()) {
            int c = iterator.next();
            if (escape) {
                switch (c) {
                    case 'n':
                        cellBuilder.append('\n');
                        break;
                    case '\\':
                        cellBuilder.append('\\');
                        break;
                    case '|':
                        cellBuilder.append('|');
                        break;
                    default:
                        // Invalid escape. We'll just ignore it.
                        cellBuilder.append("\\");
                        cellBuilder.appendCodePoint(c);
                        break;
                }
                escape = false;
            } else {
                if (c == '\\') {
                    escape = true;
                } else if (c == '|') {
                    if (beforeFirst) {
                        // Skip the first empty span
                        beforeFirst = false;
                    } else {
                        String cell = cellBuilder.toString();
                        String leftTrimmedCell = ltrim(cell);
                        int cellIndent = symbolCount(cell) - symbolCount(leftTrimmedCell);
                        lineSpans.add(new GherkinLineSpan(cellStart + cellIndent + OFFSET, rtrim(leftTrimmedCell)));
                    }
                    cellBuilder = new StringBuilder();
                    cellStart = col + 1;
                } else {
                    cellBuilder.appendCodePoint(c);
                }
            }
            col++;
        }
        return lineSpans;
    }

    @Override
    public boolean startsWithTitleKeyword(String text) {
        int textLength = text.length();
        return trimmedLineText.length() > textLength &&
                trimmedLineText.startsWith(text) &&
                trimmedLineText.substring(textLength, textLength + GherkinLanguageConstants.TITLE_KEYWORD_SEPARATOR.length())
                        .equals(GherkinLanguageConstants.TITLE_KEYWORD_SEPARATOR);
    }

}

File: gherkin/dart/lib/src/language/GherkinLine.dart
import 'package:characters/characters.dart';
import 'package:gherkin/extensions.dart';
import 'package:gherkin/exceptions.dart';
import 'package:gherkin/language.dart';

import 'GherkinLanguageConstants.dart';

class GherkinLine implements IGherkinLine
{
  // TODO: set this to 0 when/if we change to 0-indexed columns
  static final int _offset = 1;

  final String _lineText;
  final int _lineNumber;
  final String _trimmedLineText;
  final bool _isEof;

  @override
  bool get isEmpty => false;//_lineText.isEmpty && _lineNumber.isMin;

  @override
  bool get isNotEmpty => !isEmpty;

  GherkinLine(this._lineText, this._lineNumber)
    : _trimmedLineText = _lineText.trimLeft()
    , _isEof = false;

  GherkinLine.Eof()
      : _lineText = Strings.empty
      , _lineNumber = Int.min
      , _trimmedLineText = Strings.empty
      , _isEof = true;

  @override
  bool get isEof => _isEof;

  @override
  bool get isNotEof => !isEof;

  /// NOP
  @override
  void detach() {}

  int get lineNumber => _lineNumber;

  @override
  String getLineText([int indentToRemove=0]) {
    if (indentToRemove < 0 || indentToRemove > indent) {
      return _trimmedLineText;
    }
    return _lineText.substring(indentToRemove);
  }

  @override
  String getRestTrimmed(int length) =>
      _trimmedLineText.substring(length).trim();

  /// TODO Veja o fonte desta classe no Java e note que esse atributo é final
  @override
  int get indent => _lineText.length - _trimmedLineText.length;

  @override
  bool get isEmptyLine => _trimmedLineText.isEmpty;

  @override
  bool startsWith(String prefix) => _trimmedLineText.startsWith(prefix);

  @override
  bool startsWithTitleKeyword(String text) {
    var textLength = text.length;
    if( _trimmedLineText.length <= textLength ) {
      return false;
    }
    var b2 = _trimmedLineText.startsWith(text);
    var separatorLength = GherkinLanguageConstants.titleKeywordSeparator.length;
    var substr = _trimmedLineText.substring(textLength, textLength+separatorLength);
    var b3 = substr == GherkinLanguageConstants.titleKeywordSeparator;
    return b2 && b3;
  }

  @override
  Iterable<GherkinLineSpan> get tags {
    var tags = <GherkinLineSpan>[];
    final pattern = RegExp(r'\s'+GherkinLanguageConstants.commentPrefix);
    var parts = _trimmedLineText.splitWithLimit(pattern, limit:2);
    var uncommentedLine = parts[0];

    var indexInUncommentedLine = 0;
    var elements = uncommentedLine.split(GherkinLanguageConstants.tagPrefix);

    for (var element in elements) {
      var token = element.rtrim();
      if (token.isEmpty) {
        continue;
      }
      var symbolLength = uncommentedLine.substring(0, indexInUncommentedLine).length;
      var column = indent + symbolLength + 1;
      final subPattern = RegExp(r'^\S+$');
      if (!subPattern.hasMatch(token)) {
        throw ParserException('A tag may not contain whitespace'
            , Location(lineNumber, column));
      }
      tags.add(GherkinLineSpan(column, GherkinLanguageConstants.tagPrefix+token));
      indexInUncommentedLine += element.length + 1;
    }

    return tags;
  }

  /// TODO usar yield* como no C#
  @override
  Iterable<GherkinLineSpan> get tableCells {
    var lineSpans = <GherkinLineSpan>[];
    var cellBuffer = StringBuffer();
    var beforeFirst = true;
    var col = 0;
    var cellStart = 0;
    var escape = false;
    var iterator = _lineText.characters;
    for (var chr in iterator) {
      if (escape) {
        switch (chr) {
          case GherkinLanguageConstants.tableCellNewLineEscape:
            cellBuffer.write('\n');
            break;
          case GherkinLanguageConstants.tableCellEscapeChar:
            cellBuffer.write(r'\');
            break;
          case GherkinLanguageConstants.tableCellSeparator:
            cellBuffer.write('|');
            break;
          default:
          // Invalid escape. We'll just ignore it.
            cellBuffer.write(r'\');
            cellBuffer.write(chr);
            break;
        }
        escape = false;
      }
      else {
        switch (chr) {
          case GherkinLanguageConstants.tableCellEscapeChar:
            escape = true;
            break;
          case GherkinLanguageConstants.tableCellSeparator:
            if (beforeFirst) {
              // Skip the first empty span
              beforeFirst = false;
            }
            else {
              final cell = cellBuffer.toString();
              final leftTrimmedCell = cell.ltrimKeepNewLines();
              //! int cellIndent = symbolCount(cell) - symbolCount(leftTrimmedCell);
              final cellIndent = cell.length - leftTrimmedCell.length;
              final column = cellStart + cellIndent + _offset;
              final text = leftTrimmedCell.rtrimKeepNewLines();
              var lineSpan = GherkinLineSpan(column, text);
              lineSpans.add(lineSpan);
            }
            cellBuffer = StringBuffer();
            cellStart = col + 1;
            break;
          default:
            cellBuffer.write(chr);
        }
      }
      col++;
    }
    return lineSpans;
  }
}

/// Returns true if rune represents a whitespace character.
///
/// The definition of whitespace matches that used in String.trim which is based
/// on Unicode 6.2. This maybe be a different set of characters than the
/// environment's RegExp definition for whitespace, which is given by the
/// ECMAScript standard: http://ecma-international.org/ecma-262/5.1/#sec-15.10
///
/// (c) https://pub.dev/documentation/quiver/latest/quiver.strings/isWhitespace.html
bool isWhitespace(int rune) =>
    (rune >= 0x0009 && rune <= 0x000D) || // TAB, CR
        rune == 0x0020 || // SPC
        rune == 0x0085 || // Next line
        rune == 0x00A0 || // No-break space
        rune == 0x1680 ||
        rune == 0x180E ||
        (rune >= 0x2000 && rune <= 0x200A) ||
        rune == 0x2028 ||
        rune == 0x2029 ||
        rune == 0x202F ||
        rune == 0x205F ||
        rune == 0x3000 ||
        rune == 0xFEFF;

